### ISSUE TITLE
cython compilation error in master branch

### ISSUE THREAD

**Issue by thisch** [bempp/bempp#206]
Since Feb 14th my CI bempp job reports the following compilation errors.

```
[ 91%] Generating c/c++ source /tmp/bempptest/bempp/build/python_package/bempp/include/bempp/grid/entity_iterator.pyx with cython ()

Error compiling Cython file:
------------------------------------------------------------
...
cdef class EntityIterator0:

    cdef cbool finished(self):
        return deref(self.impl_).finished()

    cdef void c_next(self) except +catch_exception:
        ^
------------------------------------------------------------

/tmp/bempptest/bempp/build/python_package/bempp/include/bempp/grid/entity_iterator.pyx:22:9: Signature not compatible with previous declaration

Error compiling Cython file:
------------------------------------------------------------
...
from bempp.grid.entity_pointer cimport EntityPointer0

cdef class EntityIterator0:
    cdef unique_ptr[c_EntityIterator[codim_zero]] impl_
    cdef cbool finished(self)
    cdef void c_next(self) except *
                   ^
------------------------------------------------------------

/tmp/bempptest/bempp/build/python_package/bempp/include/bempp/grid/entity_iterator.pxd:20:20: Previous declaration is here

Error compiling Cython file:
------------------------------------------------------------
...
cdef class EntityIterator1:

    cdef cbool finished(self):
        return deref(self.impl_).finished()

    cdef void c_next(self) except +catch_exception:
        ^
------------------------------------------------------------

/tmp/bempptest/bempp/build/python_package/bempp/include/bempp/grid/entity_iterator.pyx:52:9: Signature not compatible with previous declaration

Error compiling Cython file:
------------------------------------------------------------
...
from bempp.grid.entity_pointer cimport EntityPointer1

cdef class EntityIterator1:
    cdef unique_ptr[c_EntityIterator[codim_one]] impl_
    cdef cbool finished(self)
    cdef void c_next(self) except *
                   ^
------------------------------------------------------------

/tmp/bempptest/bempp/build/python_package/bempp/include/bempp/grid/entity_iterator.pxd:29:20: Previous declaration is here

Error compiling Cython file:
------------------------------------------------------------
...
cdef class EntityIterator2:

    cdef cbool finished(self):
        return deref(self.impl_).finished()

    cdef void c_next(self) except +catch_exception:
        ^
------------------------------------------------------------

/tmp/bempptest/bempp/build/python_package/bempp/include/bempp/grid/entity_iterator.pyx:82:9: Signature not compatible with previous declaration

Error compiling Cython file:
------------------------------------------------------------
...
from bempp.grid.entity_pointer cimport EntityPointer2

cdef class EntityIterator2:
    cdef unique_ptr[c_EntityIterator[codim_two]] impl_
    cdef cbool finished(self)
    cdef void c_next(self) except *
                   ^
------------------------------------------------------------

/tmp/bempptest/bempp/build/python_package/bempp/include/bempp/grid/entity_iterator.pxd:38:20: Previous declaration is here
Linking CXX shared module ../../../python_package/bempp/grid/entity.so
```

**Comment by tbetcke** [comment]
Thanks for the info. I am using Python 0.21 and haven't noticed any issues. However, looking at the error message it seems indeed that the exception definition may cause the trouble. I will look into it.

**Comment by thisch** [comment]
Most probably this was caused by the upgrade of Cython. The last successful build was with Cython 0.21.2-py34_0.

**Comment by thisch** [comment]
I was able to reproduce this issue on a different system/machine

**Comment by thisch** [comment]
Seems to be a cython bug see
https://groups.google.com/forum/#!topic/cython-users/r3dQSfOKMjI

**Comment by scoder** [comment]
Older versions of Cython used to be too permissive here, which could easily lead to bugs (as shown by the code snippets in the error output). The compiler is stricter now and catches these.

**Comment by tbetcke** [comment]
Agree. A fix will be uploaded as part of version 2.9.1 that will come once the eigen3 feature branch is finished and merged (probably in a few days).
